man/kismet_hopper.1
.Dd July 2002
.Dt KISMET_HOPPER 1
.Os "Kismet"
.Sh NAME
.Nm kismet_hopper
.Nd Wireless sniffing and monitoring
.Sh SYNOPSIS
.Ar kismet_hopper
.Op Fl hnd
.Op Fl f Ar file
.Op Fl c Ar capture-source
.Op Fl s Ar hop-sequence
.Op Fl v Ar velocity
.Sh DESCRIPTION
kismet_hopper is a component of kismet
which changes the channel a wireless card is tuned to several times a second to
enable monitoring across all of the channels in 802.11b.
.Sh OPTIONS
.Bl -tag -width flag
.It Fl f
Use an alternate config file
.It Fl c
Override capture source lines (type,interface,name).  Refer to kismet.conf(5) for more information.
.It Fl d
Divide the channel spectrum among multiple cards
.It Fl n
Use international channels (1-14) instead of US channels (1-12)
.It Fl s
Use a custom hop sequence instead of the default.  When making a hop sequence, 
remember that ajacent channels overlap.
.It Fl v
Use a custom hop velocity (channels per second).  Many cards exhibit problems under
high hop velocities.
.It Fl h
Help
.El
.Sh SEE ALSO
kismet.conf(5), kismet_ui.conf(5), gpsmap(1), kismet(1), kismet_monitor(1)


CHANGELOG
REMEMBER:  I don't always update the changelog for newcore, check the SVN
log for current progress

Apr 20 2006  devel  Fixed BSSID protocol in kismet_server
                    Added various sources
                    Added BSSID protocol decoder to client
Apr 07 2006  devel  Added server picker window
                    Added "add source" window
Mar 20 2006  devel  Added keepalive to drone protocol
Mar 12 2006  devel  I know it's been a while, stuff going on in life, sorry.
                    Added CARD protocol support to kp_frontend
                    Changed behavior of kp_Scrollable_Table to add a row
                     during a replace of a nonexistent row
Feb 22 2006  devel  Lots of un-changelog'd work in svn
                    Increased max iwpriv ioctls per interface, will produce
                     a different error on madwifi-ng using madwifi-old sources
                    Finished cli matrix widget
                    Added ModalAlert panel
                    "File->Connect" actually connects to the Kismet server
Feb 02 2006  devel  More work on panels client
                    Work on new tcp client code
                    Fixed stupid error in server that didn't send capabilities
Jan 16 2006  devel  Finished client panel single input line widget
                    Fixed menu not eating input when selecting nonexistent
                     menus or items via keyboard
                    Fixed consecutive disabled items in menus
                    Fixed stupid error in window positioning
                    Added button widget
                    Added "Connect" demo window
Jan 12 2006  devel  Finished client panel menu widget code
                    Finished client panel arbitrary text scroller widget code
                    Added Madwifi-NG support (with dyanamic interface
                     creation and destruction support)
                    Removed prism2 header length validation, not needed and
                     some drivers are broken
                    Split widgets and windows into their own files
                    Added client panel 2-field widget
                    Added picking list type demo
                    Added inline text style formatting class, added to
                     freetext widget
Jan 08 2006  devel  Started newcore client development (Current client code
                     is demo only and will not do anything useful)
                    Changed menu drivers to go to menubar mode and allow
                     picking the menu by underlined key
Dec 29 2005  devel  Added runstate loaders to kismet_server (do not use yet)
                    Added runstate support to gpsxml
                    Added runstate logging to netracker
                    Added runstate loading to netracker
                    Added runstate net totals logging/loading
                    Runstate loading should be usable
Dec 28 2005  devel  Added runstate dumpfile framework for freezing the Kismet
                     running state
                    Tweaked runstate dumpfile parsing
                    Added bcm43xx source
                    Started adding dumpfile resume code
Dec 21 2005  devel  Fixed error in netframe caching
Dec 19 2005  devel  Added dynamic source creation/removal via ADDSOURCE and
                     DELSOURCE commands
Dec 18 2005  devel  Fixed drone starting if drone server doesn't start
                    Added --source-options command line switch
Dec 16 2005  devel  Fixed vector resize with uclibc++
                    Added uclibc++ autodetect in autoconf
                    Added libm autodetect in autoconf
                    Removed const qualifier in macaddr.h to allow compiling on
                     uclibc c++
Dec 15 2005  devel  Added remote drone name to messages passed from it
                    Fixed drone stream syncing problem which caused sentinel
                     errors
                    Fixed compilation error when libpcap disabled
Dec 14 2005  devel  Fixed queuing of dumpfiles when dumping is disabled
Dec 13 2005  devel  Fixed loading of servername in server/drone
                    Restored wrt54prism capture source for OpenWRT
                    Fixed packet processing loop in packetsource_drone
Dec 12 2005  devel  Revamped AddSource callback API into a generic SourceAction
                     API
                    Wrote drone interface channel push code
                    Added SourceAction hooks for setting hop, vectors
                    Added drone server-side handling of channelset for hopping
                     and vector
                    Changed internal channel representation to unsigned int
                    Added packet sending to drone client
                    Added channelset packet generation to drone client
                    Added proper bitmap handling of channelset frames in
                     drone server
                    Added channel sequence command (CHANSEQ)
                    Fixed card commands (hop, etc) from segfaulting
                    Fixed dumpfiles unregistering callbacks during destruction
                    Fixed more blatant errors in uuid
                    Fixed stupid errors in drone server handling
Dec 11 2005  devel  Added fix to ringbuf from Shane Schisler
                    Added drone_source_packet generators to drone protocol to
                     push a capture source upstream
                    Added endian-flipping to bitmap fields in drone protocol
                    Completed server side of drone interface push system
                    Fixed UUID '<' operator
                    Added remote source pushing via drone protocol, remote sources
                     now show up
                    Added purging of virtual sources when drone connection is
                     broken
Dec 10 2005  devel  Added UUID to drone protocol

kismet-devel/kismet_hopper.cc
/*
    This file is part of Kismet

    Kismet is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    Kismet is distributed in the hope that it will be useful,
      but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Kismet; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <signal.h>
#include <vector>
#include "getopt.h"
#include "configfile.h"

#ifndef exec_name
char *exec_name;
#endif

typedef struct capturesource {
    char *interface;
    int chanpos;
    char *cardtype;
    const char *cmd_template;
};

#define prism2 "wlanctl-ng %s lnxreq_wlansniff channel=%d enable=true"
#define prism2_pcap "wlanctl-ng %s lnxreq_wlansniff channel=%d enable=true prismheader=true"
#define prism2_bsd "prism2ctl %s -f %d"
#define prism2_hostap "iwconfig %s channel %d"
#define orinoco "iwpriv %s monitor 1 %d"

// This one is nonstandard, but then, the wsp100 is in general.  We push channelhop, channel mask,
// and hop speed then stop
#define wsp100 "snmpset -c public %s .1.3.6.1.4.1.14422.1.3.13 i 1 .1.3.6.1.4.1.14422.1.3.15 i %d .1.3.6.1.4.1.14422.1.3.14 i %ld"

#define pidpath "/var/run/kismet_hopper.pid"
#define conpath "/tmp/kismet_hopper.control"

const char *config_base = "kismet.conf";

// Channel rotations to maximize hopping for US and international frequencies
int us_channels[] = {1, 6, 11, 2, 7, 3, 8, 4, 9, 5, 10, -1};
int intl_channels[] = {1, 7, 13, 2, 8, 3, 14, 9, 4, 10, 5, 11, 6, 12, -1};

int *parse_hopseq(char* param) {
    int *result;
    char *pos;
    char *nextpos;

    vector<int> assemble;

    //    result = (int*)malloc(sizeof(int)*15);

    pos = param;

    int res;

    while ((nextpos=strchr(pos,',')) != NULL) {
        *nextpos = 0;
        nextpos++;

        if (sscanf(pos, "%d", &res) < 1) {
            fprintf(stderr, "Illegal custom channel sequence %s\n", param);
            exit(1);
        }

        assemble.push_back(res);
        pos = nextpos;
    }

    // Catch the last one
    if (sscanf(pos, "%d", &res) < 1) {
        fprintf(stderr, "Illegal custom channel sequence %s\n", param);
        exit(1);
    }
    assemble.push_back(res);

    result = (int *) malloc(sizeof(int) * (assemble.size() + 1));

    for (unsigned int x = 0; x < assemble.size(); x++) {
        if (assemble[x] < 1 || assemble[x] > 14) {
            fprintf(stderr, "Illegal custom channel %d - valid channels are 1-14.\n",
                    assemble[x]);
            exit(1);
        }
        result[x] = assemble[x];
    }

    result[assemble.size()] = -1;

    return result;
}

int Usage(char *argv) {
    printf("Usage: %s [OPTION]\n", argv);
    printf(
           "  -f, --config-file <file>     Use alternate config file\n"
           "  -d, --divide-channels        Divide channels across multiple cards when possible\n"
           "  -c, --capture-source <src>   Packet capture source line (type,interface,name)\n"
           "  -n, --international          Use international channels (1-14)\n"
           "  -s, --hopsequence            Use given hop sequence\n"
           "  -v, --velocity               Hopping velocity (hops per second)\n"
           "  -h, --help                   What do you think you're reading?\n");
    exit(1);
}

void CatchShutdown(int sig) {
    fprintf(stderr, "kismet_hopper shutting down.\n");
    unlink(pidpath);
    exit(0);
}

int main(int argc, char *argv[]) {
    int *chanlist;
    FILE *pgsock;
    char *configfile = NULL;
    char *channame = "United States";
    struct stat fstat;

    FILE *pidfile;
    int pid;

    exec_name = argv[0];

    unsigned long interval;
    int freq = 3;

    if (stat(pidpath, &fstat) == 0) {
        fprintf(stderr, "Detected pid file '%s'.  Make sure another instance of kismet_hopper\n"
                "isn't running, and remove this file.", pidpath);
        CatchShutdown(-1);
    }

    static struct option long_options[] = {   /* options table */
        { "config-file", required_argument, 0, 'f' },
        { "international", no_argument, 0, 'n' },
        { "capture-source", required_argument, 0, 'c' },
        { "hopsequence", required_argument, 0, 's' },
        { "velocity", required_argument, 0, 'v' },
        { "divide-channels", no_argument, 0, 'd' },
        { "help", no_argument, 0, 'h' },
        { 0, 0, 0, 0}
    };

    vector<string> source_input_vec;
    int divide = 0;
    int divisions = 0;

    chanlist = us_channels;
    int option_index;
    while(1) {
        int r = getopt_long(argc, argv, "fnc:s:v:hd",
                            long_options, &option_index);
        if (r < 0) break;

        switch(r) {
        case 'f':
            configfile = optarg;
            break;
        case 'n':
            chanlist = intl_channels;
	    channame = "International";
            break;
        case 'v':
            if (sscanf(optarg, "%d", &freq) != 1) {
                fprintf(stderr, "Invalid number for channels-per-second\n");
                Usage(argv[0]);
            }
            break;
	case 's':
            chanlist = parse_hopseq(optarg);
            channame = "Custom channels";
            if (!chanlist)
                exit(1);
            break;
        case 'c':
            // Capture type
            source_input_vec.push_back(optarg);
            break;
        case 'd':
            // Divide channels
            divide = 1;
            break;
        default:
            Usage(argv[0]);
            break;
        }
    }

    ConfigFile *conf = new ConfigFile;

    // If we haven't gotten a command line config option...
    if (configfile == NULL) {
        configfile = (char *) malloc(1024*sizeof(char));
        snprintf(configfile, 1024, "%s/%s", SYSCONF_LOC, config_base);
    }

    // Parse the config and load all the values from it and/or our command
    // line options.  This is a little soupy but it does the trick.
    if (conf->ParseConfig(configfile) < 0) {
        exit(1);
    }

    // Catch old configs and yell about them
    if (conf->FetchOpt("cardtype") != "" || conf->FetchOpt("captype") != "" ||
        conf->FetchOpt("capinterface") != "") {
        fprintf(stderr, "FATAL:  Old config file options found.  To support multiple sources, Kismet now\n"
                "uses a new config file format.  Please consult the example config file in your Kismet\n"
                "source directory, OR do 'make forceinstall' and reconfigure Kismet.\n");
        exit(1);
    }

    // Read all of our packet sources, tokenize the input and then start opening
    // them.

    vector<capturesource *> packet_sources;

    string sourceopt;

    if (freq == 0) {
        fprintf(stderr, "No point in hopping 0 channels.  Setting velocity to 1.\n");
        freq = 3;
    } else if (freq > 10) {
        if (freq > 100) {
            fprintf(stderr, "Cannot hop more than 100 channels per second, setting velocity to 100\n");
            freq = 100;
        } else {
            fprintf(stderr, "WARNING: Velocities over 10 are not reccomended and may cause problems.\n");
        }
    }

    interval = 1000000 / freq;
    fprintf(stderr, "Hopping %d channel%sper second (%ld microseconds per channel)\n",
            freq, freq > 1 ? "s " : " ", interval);

    // Read the config file if we didn't get any sources on the command line
    if (source_input_vec.size() == 0) {
        int sourcenum = 0;
        char sourcestr[16];

        while (snprintf(sourcestr, 16, "source%d", sourcenum++) &&
               (sourceopt = conf->FetchOpt(sourcestr)) != "") {
            source_input_vec.push_back(sourceopt);
        }
    }

    if (source_input_vec.size() == 0) {
        fprintf(stderr, "FATAL:  No valid packet sources defined in config or passed on command line.\n");
        exit(1);
    }

    // Now tokenize the sources
    for (unsigned int x = 0; x < source_input_vec.size(); x++) {
        sourceopt = source_input_vec[x];

        unsigned int begin = 0;
        unsigned int end = sourceopt.find(",");
        vector<string> optlist;

        while (end != string::npos) {
            string subopt = sourceopt.substr(begin, end-begin);
            begin = end+1;
            end = sourceopt.find(",", begin);
            optlist.push_back(subopt);
        }
        optlist.push_back(sourceopt.substr(begin, sourceopt.size() - begin));

        if (optlist.size() < 3) {
            fprintf(stderr, "FATAL:  Invalid source line '%s'\n", sourceopt.c_str());
            exit(1);
        }

        capturesource *newsource = new capturesource;
        newsource->cardtype = strdup(optlist[0].c_str());
        newsource->interface = strdup(optlist[1].c_str());
        newsource->chanpos = 0;
        packet_sources.push_back(newsource);
        optlist.clear();
    }

    for (unsigned int src = 0; src < packet_sources.size(); src++) {
        char *type = packet_sources[src]->cardtype;

        if ((!strcasecmp(type, "cisco") || !strcasecmp(type, "cisco_bsd")) &&
            packet_sources.size() == 1) {
            fprintf(stderr, "FATAL:  Cisco cards don't need to channel hop.\n");
            exit(1);
        } else if (!strcasecmp(type, "prism2")) {
            packet_sources[src]->cmd_template = prism2;
            divisions++;
        } else if (!strcasecmp(type, "prism2_pcap")) {
            packet_sources[src]->cmd_template = prism2_pcap;
            divisions++;
        } else if (!strcasecmp(type, "prism2_bsd")) {
            packet_sources[src]->cmd_template = prism2_bsd;
            divisions++;
        } else if (!strcasecmp(type, "prism2_hostap")) {
            packet_sources[src]->cmd_template = prism2_hostap;
            divisions++;
        } else if (!strcasecmp(type, "orinoco")) {
            packet_sources[src]->cmd_template = orinoco;
            divisions++;
        } else if (!strcasecmp(type, "wsp100")) {
            // We just process this one immediately - it's annoying that it's a special
            // case, but it's so different from anything else...

            fprintf(stderr, "NOTICE:  Source %d: Setting wsp100 hopping immediately.\n", src);

            // Convert the channels into a bitmap
            int chans = 0;
            int chind = 0;
            while (chanlist[chind] != -1)
                chans |= ( 1 << chind++ );

            char cmd[1024];
            char iface[64];

            if (sscanf(packet_sources[src]->interface, "%64[^:]:", iface) < 1) {
                fprintf(stderr, "FATAL:  Source %d: Could not parse host from interface.\n", src);
                exit(0);
            }

            snprintf(cmd, 1024, wsp100, iface, chans, (interval/100));

            FILE *pgsock;

            if ((pgsock = popen(cmd, "r")) < 0) {
                fprintf(stderr, "Fatal:  Source %d: Could not popen ``%s''.  Aborting.\n", src, cmd);
                exit(0);
            }

            pclose(pgsock);

            // Exit if we only have this one thing
            if (source_input_vec.size() == 1) {
                fprintf(stderr, "NOTICE:  Terminating hopper after setting hop modes on wsp100.  No other devices.\n");
                exit(0);
            }

            packet_sources[src]->cmd_template = NULL;
        } else {
            fprintf(stderr, "FATAL: Source %d: Unknown card type '%s'.\n", src, type);
            exit(1);
        }

        free(type);
        source_input_vec.clear();
    }

    // Free up the memory
    delete conf;
    conf = NULL;

    if ((pidfile = fopen(pidpath, "w")) == NULL) {
        fprintf(stderr, "FATAL:  Could not open PID file '%s' for writing.\n", pidpath);
        exit(1);
    }

    pid = getpid();
    fprintf(pidfile, "%d\n", pid);
    fclose(pidfile);

    signal(SIGINT, CatchShutdown);
    signal(SIGTERM, CatchShutdown);
    signal(SIGHUP, CatchShutdown);
    signal(SIGPIPE, SIG_IGN);

    if (divide && divisions > 1) {
        int nchannels = 0;
        while (chanlist[nchannels++] != -1) ;
        fprintf(stderr, "Dividing %d channels into %d segments.\n", nchannels, divisions);
        for (unsigned int ofst = 0; ofst < packet_sources.size(); ofst++) {
            packet_sources[ofst]->chanpos = (nchannels / divisions) * ofst;
            printf("Source %d starting at offset %d\n", ofst, packet_sources[ofst]->chanpos);
        }
    }

    for (unsigned int src = 0; src < packet_sources.size(); src++)
        fprintf(stderr, "%s - Source %d - Channel hopping (%s) on interface %s\n",
                argv[0], src, channame, packet_sources[src]->interface);

    char cmd[1024];
    int statime = 0;
    while (1) {
        if (statime++ > 5) {
            statime = 0;
            if (stat(conpath, &fstat) == 0) {
                fprintf(stderr, "Detected killfile %s\n", conpath);
                CatchShutdown(-1);
            }
        }

        for (unsigned int src = 0; src < packet_sources.size(); src++) {
            if (packet_sources[src] == NULL)
                continue;

            capturesource *csrc = packet_sources[src];
            snprintf(cmd, 1024, csrc->cmd_template, csrc->interface, chanlist[csrc->chanpos++]);

            if (chanlist[csrc->chanpos] == -1)
                csrc->chanpos = 0;

            if ((pgsock = popen(cmd, "r")) < 0) {
                fprintf(stderr, "FATAL:  Source %d: Could not popen ``%s''.  Aborting.\n", src, cmd);
                exit(0);
            }

            pclose(pgsock);
        }

        usleep(interval);

    }

}
